Workflow and Pseudocode for email automation system.**

<img src="static/Flowcharts/Email Workflow.jpg">

### <span style="color: green"> **Pseudocode for email automation - the full copies are inside the MS Word documents!** </span>

In [26]:
""" 
<p>Dear @{triggerOutputs()?['body/PassengerName']},</p>

<p>We regret to inform you that your flight <strong>@{triggerOutputs()?['body/FlightNumber']}</strong> from <strong>@{triggerOutputs()?['body/OriginAirport']}</strong> 
to <strong>@{triggerOutputs()?['body/DestinationAirport']}</strong>, originally scheduled to depart at <strong>@{triggerOutputs()?['body/ScheduledDepartureTime']}</strong>, 
has been delayed by <strong>@{triggerOutputs()?['body/DelayDuration']}</strong> due to <strong>@{triggerOutputs()?['body/ReasonForDelay']}</strong>.</p>

<p>We sincerely apologize for any inconvenience this may cause and appreciate your patience as we work to resolve the situation.</p>

<h3>What You Need to Know:</h3>
<ul>
    <li><strong>New Estimated Departure Time:</strong> @{triggerOutputs()?['body/UpdatedDepartureTime']}</li>
    <li><strong>Current Status:</strong> @{triggerOutputs()?['body/FlightStatus']}</li>
    <li><strong>Gate Information:</strong> @{triggerOutputs()?['body/GateInfo']}</li>
</ul>

<h3>Compensation & Support:</h3>
<p>As this delay exceeds <strong>@{triggerOutputs()?['body/ThresholdDuration']}</strong> or follows repeated delays, we would like to offer you the following:</p>
<p><strong>@{triggerOutputs()?['body/CompensationDetails']}</strong></p>
<p>To claim your compensation or discuss alternative arrangements, please visit our <a href="@{triggerOutputs()?['body/CompensationPortalLink']}">Compensation Claim Portal</a> 
or contact our customer support team.</p>

<h3>What You Can Do Now:</h3>
<ul>
    <li><strong>Check Flight Updates:</strong> Stay updated via <a href="@{triggerOutputs()?['body/AirlineAppLink']}">our website or app</a>.</li>
    <li><strong>Rebook Your Flight:</strong> Manage your booking at <a href="@{triggerOutputs()?['body/ManageBookingLink']}">this link</a>.</li>
    <li><strong>Speak to a Representative:</strong> Email us at 
    <a href="mailto:@{triggerOutputs()?['body/CustomerServiceEmail']}">@{triggerOutputs()?['body/CustomerServiceEmail']}</a> 
    or call <strong>@{triggerOutputs()?['body/CustomerSupportPhone']}</strong>.</li>
</ul>

<p>We sincerely appreciate your understanding and patience. Your comfort and convenience are our top priority, 
and we are doing everything we can to minimize the impact of this delay.</p>

<p>Thank you for choosing <strong>@{triggerOutputs()?['body/AirlineName']}</strong>. We look forward to welcoming you on board soon.</p>

<p>Best regards,<br>
@{triggerOutputs()?['body/CustomerSupportTeam']}<br>
@{triggerOutputs()?['body/AirlineName']}<br>
<a href="@{triggerOutputs()?['body/AirlineWebsite']}">@{triggerOutputs()?['body/AirlineWebsite']}</a></p>
"""

' \n<p>Dear @{triggerOutputs()?[\'body/PassengerName\']},</p>\n\n<p>We regret to inform you that your flight <strong>@{triggerOutputs()?[\'body/FlightNumber\']}</strong> from <strong>@{triggerOutputs()?[\'body/OriginAirport\']}</strong> \nto <strong>@{triggerOutputs()?[\'body/DestinationAirport\']}</strong>, originally scheduled to depart at <strong>@{triggerOutputs()?[\'body/ScheduledDepartureTime\']}</strong>, \nhas been delayed by <strong>@{triggerOutputs()?[\'body/DelayDuration\']}</strong> due to <strong>@{triggerOutputs()?[\'body/ReasonForDelay\']}</strong>.</p>\n\n<p>We sincerely apologize for any inconvenience this may cause and appreciate your patience as we work to resolve the situation.</p>\n\n<h3>What You Need to Know:</h3>\n<ul>\n    <li><strong>New Estimated Departure Time:</strong> @{triggerOutputs()?[\'body/UpdatedDepartureTime\']}</li>\n    <li><strong>Current Status:</strong> @{triggerOutputs()?[\'body/FlightStatus\']}</li>\n    <li><strong>Gate Information:</strong> 

### <span style="color: green"> Alternative: AI-based copy generation </span>

In [27]:
## COPY  #1
"""
Professional and empathetic email template for passengers affected by a delayed flight, incorporating compensation details for repeated or very long delays, 
customer service contact information, and a clear call to action.
Remember that three number signs “###” means that you should input text, following the guidelines and keeping a formal tone-of-voice.

This is the structure you must follow:

Subject: Important Update: Your Flight [Flight Number] - Delay Notification
Dear [Passenger Name],
### text ###
What You Need to Know:
### text ###
Compensation & Support:
### text ###
We sincerely appreciate your understanding and patience… ###  follow with more text ### 
Best regards,
[Your Airline’s Customer Support Team]
[Airline Name]
[Airline Website]
-- End of Message --
"""

## COPY  #2
"""
The template you are going to generate must ensure clarity, provide actionable next steps, and include an empathetic tone while maintaining a professional approach. 
Professional and empathetic email template for passengers affected by a delayed flight, incorporating compensation details for repeated or very long delays, 
customer service contact information, and a clear call to action. You should structure the email as a second reachout to the customer, sent by the customer service department. 
The aim is to create engagement and fidelization.
Remember that three number signs “###” means that you should input text, following the guidelines and keeping a formal tone-of-voice.

This is the structure you must follow:

Subject: Flight [Flight Number] from [Origin] to [Destination] – Delay Notification and Assistance
Dear [Passenger Name],
### text ###
Flight Details
### text ###
Next Steps for Passengers
### text ###
Compensation and Support
### text ###
Customer Service Contact Information
### text ###
Useful Links & Resources
### text ###
Again, we truly apologize for the inconvenience this delay has caused. Thank you for your patience and understanding in this matter. We value your trust in [Airline Name], 
and we are fully committed to getting you to [Destination] as safely and quickly as possible.
Sincerely,
[Airline Name] Customer Care Team
-- End of Message –
"""

'\nThe template you are going to generate must ensure clarity, provide actionable next steps, and include an empathetic tone while maintaining a professional approach. \nProfessional and empathetic email template for passengers affected by a delayed flight, incorporating compensation details for repeated or very long delays, \ncustomer service contact information, and a clear call to action. You should structure the email as a second reachout to the customer, sent by the customer service department. \nThe aim is to create engagement and fidelization.\nRemember that three number signs “###” means that you should input text, following the guidelines and keeping a formal tone-of-voice.\n\nThis is the structure you must follow:\n\nSubject: Flight [Flight Number] from [Origin] to [Destination] – Delay Notification and Assistance\nDear [Passenger Name],\n### text ###\nFlight Details\n### text ###\nNext Steps for Passengers\n### text ###\nCompensation and Support\n### text ###\nCustomer Servi

------------------

-------------------------

<img src="static/Flowcharts/Coding Flow.jpg">

In [28]:
import pandas as pd 
import pyodbc
import seaborn as sb
import plotly.express as px
from copy import deepcopy

pd.set_option('display.max_columns', None)

import warnings
warnings.filterwarnings("ignore")

import plotly.offline as pyo
pyo.init_notebook_mode(connected=True)


data = pd.read_csv("data.csv")

## EDA - Exploratory Data Analysis

In [29]:
data.describe()

AIR_SYSTEM_DELAY     AIR_TIME  AIRLINE_DELAY  ARRIVAL_DELAY  \
count        421.000000  2500.000000     421.000000    2500.000000   
mean          13.581948   115.029200      19.600950       2.953600   
std           24.236161    74.300184      48.401126      36.374363   
min            0.000000    16.000000       0.000000     -53.000000   
25%            0.000000    61.000000       0.000000     -13.250000   
50%            2.000000    94.000000       4.000000      -6.000000   
75%           18.000000   145.000000      20.000000       6.000000   
max          202.000000   629.000000     651.000000     651.000000   

       ARRIVAL_TIME    CANCELLED          DAY  DAY_OF_WEEK  DEPARTURE_DELAY  \
count   2508.000000  2555.000000  2555.000000  2555.000000      2510.000000   
mean    1451.973684     0.018004    15.720548     4.000000         8.014741   
std      533.325852     0.132991     8.797969     1.997649        34.351049   
min        1.000000     0.000000     1.000000     1.000000       -26.000000   
25%     1031.000000     0.000000     8.000000     2.000000        -5.000000   
50%     1445.000000     0.000000    16.000000     4.000000        -2.000000   
75%     1907.250000     0.000000    23.000000     6.000000         5.000000   
max     2400.000000     1.000000    31.000000     7.000000       665.000000   

       DEPARTURE_TIME     DISTANCE     DIVERTED  ELAPSED_TIME  FLIGHT_NUMBER  \
count     2510.000000  2555.000000  2555.000000   2500.000000    2555.000000   
mean      1298.406773   836.065362     0.003523    138.452000    2130.446184   
std        512.952110   629.966122     0.059258     75.964423    1754.366191   
min          1.000000    67.000000     0.000000     32.000000       2.000000   
25%        853.250000   374.000000     0.000000     84.000000     690.000000   
50%       1257.500000   650.000000     0.000000    117.500000    1640.000000   
75%       1728.750000  1072.000000     0.000000    170.000000    3160.500000   
max       2400.000000  4983.000000     1.000000    646.000000    7423.000000   

       LATE_AIRCRAFT_DELAY        MONTH  SCHEDULED_ARRIVAL  \
count            421.00000  2555.000000        2555.000000   
mean              23.11639     6.526027        1456.465362   
std               42.13638     3.448526         519.888474   
min                0.00000     1.000000           1.000000   
25%                0.00000     4.000000        1035.000000   
50%                1.00000     7.000000        1450.000000   
75%               29.00000    10.000000        1900.000000   
max              296.00000    12.000000        2359.000000   

       SCHEDULED_DEPARTURE  SCHEDULED_TIME  SECURITY_DELAY      TAXI_IN  \
count          2555.000000     2555.000000      421.000000  2508.000000   
mean           1295.328767      143.007436        0.052257     7.451754   
std             500.462061       76.678944        1.072214     5.994718   
min               2.000000       30.000000        0.000000     1.000000   
25%             855.000000       86.000000        0.000000     4.000000   
50%            1259.000000      122.000000        0.000000     6.000000   
75%            1720.000000      175.000000        0.000000     9.000000   
max            2359.000000      680.000000       22.000000   135.000000   

          TAXI_OUT  WEATHER_DELAY    WHEELS_ON   WHEELS_OFF    YEAR  
count  2509.000000     421.000000  2508.000000  2509.000000  2555.0  
mean     15.972499       1.648456  1445.797847  1318.001993  2015.0  
std       8.612866      14.951507   530.366546   514.766934     0.0  
min       3.000000       0.000000     1.000000     2.000000  2015.0  
25%      11.000000       0.000000  1027.750000   908.000000  2015.0  
50%      14.000000       0.000000  1440.500000  1312.000000  2015.0  
75%      19.000000       0.000000  1904.000000  1740.000000  2015.0  
max     109.000000     285.000000  2400.000000  2350.000000  2015.0

<span style="color: green"> 

### We can observe that the **average Air System delay (AS)** is relatively low at **13.58 minutes**, as is the **Airline delay (AL) at 19.60 minutes**, **Arrival delay at 2.95 minutes**, **Late Aircraft delay at 23.11 minutes**, **Security delay at 0.05 minutes**, and **Weather delay at 1.64 minutes**. However, these values are **strongly skewed by non-delayed flights**, so we will conduct a deeper analysis later by focusing specifically on the delayed flight subset.  

### At this stage, we can note that **airline-related delays appear to be the most frequent, along with weather-related delays**, suggesting a possible correlation. We will plot their correlation later to assess the degree of causality between these factors.  

### Additionally, we see that the **75th percentile is at 18.0 minutes**, indicating that in our dataset of approximately **2,500 flights**, there are **not many significantly long delays**. Further investigation will help clarify delay distribution and outliers.  

</span>

In [30]:
data.isnull().sum().sort_values(ascending=False).head(10)

CANCELLATION_REASON    2509
AIR_SYSTEM_DELAY       2134
SECURITY_DELAY         2134
LATE_AIRCRAFT_DELAY    2134
AIRLINE_DELAY          2134
WEATHER_DELAY          2134
ARRIVAL_DELAY            55
ELAPSED_TIME             55
AIR_TIME                 55
TAXI_IN                  47
dtype: int64

<span style="color: green">

### The consumer **primarily** focuses on arrival delay, as their main concern is reaching their destination on time. <br>

### However, other types of delays provide valuable insights for optimizing operational efficiency and enhancing customer experience. For example, while passengers don’t want to arrive late, they also don’t want to spend two extra hours sitting in the cabin before takeoff. Even if the aircraft compensates for the delay in-flight and arrives on time, the overall experience may still be frustrating. <br>

### Another example is that consistently departing late could indicate inefficiencies in the boarding process or fueling times. Repeated delays at departure may not always impact arrival times if the flight makes up for lost time in the air, but they signal operational bottlenecks that could be optimized for better efficiency and passenger experience. <br>

#### Let's have a look!
</span>

In [31]:
data_reduced = deepcopy(data[['AIR_SYSTEM_DELAY', 'AIRLINE_DELAY',
       'ARRIVAL_DELAY', 'Delay Category', 'DEPARTURE_DELAY',
       'LATE_AIRCRAFT_DELAY', 'SECURITY_DELAY', 'WEATHER_DELAY']]).fillna(0)

value_vars=[
       'AIR_SYSTEM_DELAY', 'AIRLINE_DELAY',
       'ARRIVAL_DELAY', 'DEPARTURE_DELAY',
       'LATE_AIRCRAFT_DELAY', 'SECURITY_DELAY', 'WEATHER_DELAY'
       ]


data_reduced['Delay Category'].unique()

def map_delay(row, delay_type):
    if row[delay_type] == 0:
        return 'On Time'
    elif row[delay_type] <= 15:             
        return 'Short Delay <= 15 min'
    elif row[delay_type] <= 60:
        return 'Moderate Delay 15-60 min'
    elif row[delay_type] <= 240:
        return 'Long Delay 60 - 240 min'
    else:
        return 'Very Long Delay'
    

for delay_type in value_vars:
    df = deepcopy(data_reduced[[delay_type]])
    df['delay_type_mapped'] = data_reduced[[delay_type]].apply(lambda row: map_delay(row, delay_type), axis=1)
    df_adj = df[df['delay_type_mapped']  != 'On Time']
    grouped = df_adj.groupby(['delay_type_mapped', delay_type])[delay_type].value_counts().reset_index()
    graph = px.scatter(data_frame=grouped, x=delay_type, y='count', color='delay_type_mapped' ,height=300, width=1400, title=delay_type)
    pyo.iplot(graph)

<span style='color: green'>

### No anomalies detected overall. As expected, most delays fall into the **short or moderate delay** categories.

### An interesting observation is that some delay values are **negative**, which likely indicates cases where the flight **departed or arrived earlier than scheduled** rather than experiencing an actual delay. <br>

### **Bonus Insight:**  
- **Security delays** remain consistently **on time**, posing no operational concerns.  
- **Weather-related delays** are, as expected, a factor. However, since weather conditions are **uncontrollable and highly volatile**, we won’t consider them as a key variable in this analysis.  

### **Departure vs. Arrival Delay:**  
Departure and arrival delays **show a very similar distribution**, which makes sense:  
- **Departing late generally leads to arriving late.**  
- However, in some cases, the aircraft can **compensate by flying faster**, allowing passengers to **still arrive on time**.  
- Despite these adjustments, the **cause-effect relationship** remains **straightforward**—a late departure **often translates into a late arrival**.
</span>

In [37]:
no_delays_df = data[data['Delay Category'] != 'On Time'][['MONTH', 'AIRLINE', 'ARRIVAL_DELAY']]
no_delays_pivot = no_delays_df.pivot_table(index=['MONTH', 'AIRLINE'], aggfunc='count').reset_index()

airlines = data['AIRLINE'].unique()

linegraph = px.line(data_frame=no_delays_pivot, x='MONTH', y='ARRIVAL_DELAY', color='AIRLINE', height=800, title="MoM delays general outlook", facet_col='AIRLINE', facet_col_wrap=5, text='ARRIVAL_DELAY')
pyo.iplot(linegraph)

<span style="color: green">

### **Monthly Trend Analysis (MoM)**  

This graph illustrates the **Month-over-Month (MoM) trend**, highlighting the **poor performance of WN Airlines going at its ATH in December** or the **strong and constant performance of VX and HA** or that **US probably is out of business**. While this provides a general overview, a deeper dive reveals that **UA struggled significantly during the summer months**.  

#### **Key Insights & Next Steps**  
This type of insight opens the door for more detailed analysis. For instance, we could investigate **why UA experienced a higher number of delays during the summer season**.  

Potential contributing factors include:  
- **Weather conditions** – Increased delays due to seasonal storms or extreme temperatures.  
- **Operational bottlenecks** – High demand for flights while maintaining a **constant workforce** may have led to inefficiencies in scheduling, crew availability, or turnaround times.  
- **Infrastructure limitations** – Increased airport congestion or maintenance backlogs affecting punctuality.  

Understanding these factors can help refine **delay mitigation strategies** and improve airline performance.
</span>

-----------------------------

---------------------------

## Origin Airports with the **Longest Average** Departure and Arrival Delays in 2015

In [33]:
origin_data = deepcopy(data[['ORIGIN_AIRPORT', 'DEPARTURE_DELAY']])
top_10_delays_origin = origin_data.pivot_table(index='ORIGIN_AIRPORT', values='DEPARTURE_DELAY', aggfunc='mean').sort_values(by='DEPARTURE_DELAY', ascending=False).reset_index()[:10]
top_10_delays_origin

ORIGIN_AIRPORT  DEPARTURE_DELAY
0            GRK            263.0
1            PSC            245.0
2            HYA            234.0
3          13244            113.5
4          11057             74.5
5            CRP             64.0
6            ACY             51.0
7          13158             46.0
8            CRW             43.5
9            MLI             39.5

<span style="color: green">

### Several factors can contribute to **major average delays** at certain origin airports:

#### 1. **High Traffic Volume & Congestion** – Airports handling a large number of flights may experience congestion, causing frequent delays.
#### 2. **Weather Conditions** – Airports in regions with frequent storms, snow, or fog tend to suffer more delays.
#### 3. **Air Traffic Control Restrictions** – Airports in busy air corridors may be affected by air traffic control slowdowns due to limited airspace capacity.
#### 4. **International vs. Domestic Flights** – Airports with a higher proportion of long-haul international flights may experience more delays due to stricter security checks, customs procedures, and turnaround time constraints.
#### 5. **Time of Day & Peak Periods** – Airports that experience a concentration of departures during specific hours may suffer from bottlenecks, leading to longer delays.
</span>

## Origin Airports with the **Lowest Average** Departure and Arrival Delays in 2015
#### These are the ones that, on average, depart early

In [34]:
top_10_performances_origin = origin_data.pivot_table(index='ORIGIN_AIRPORT', values='DEPARTURE_DELAY', aggfunc='mean').sort_values(by='DEPARTURE_DELAY', ascending=False).reset_index()[-10:]
top_10_performances_origin

ORIGIN_AIRPORT  DEPARTURE_DELAY
279          12280            -12.0
280          12953            -12.5
281          15370            -13.0
282            BQN            -14.0
283          14262            -15.0
284          13970            -15.0
285          11049            -16.0
286            ISN            -17.0
287            GUC            -19.0
288            IAG            -26.0

---------------------------

---------------------------

In [35]:
origin_data_2 = deepcopy(data[['DESTINATION_AIRPORT', 'ARRIVAL_DELAY']])
top_10_delays_arrival = origin_data_2.pivot_table(index='DESTINATION_AIRPORT', values='ARRIVAL_DELAY', aggfunc='mean').sort_values(by='ARRIVAL_DELAY', ascending=False).reset_index()[:10]
top_10_delays_arrival

DESTINATION_AIRPORT  ARRIVAL_DELAY
0                 SAF     168.500000
1                 ISP      82.000000
2                 BZN      68.000000
3               12951      67.000000
4                 SBN      64.000000
5                 TTN      62.000000
6                 FWA      56.000000
7               12266      47.714286
8                 SYR      37.000000
9                 JAX      34.000000

<span style="color: green">

### Several factors can contribute to **major average delays** at certain origin airports:

#### 1. **International Arrival Protocols** – Flights arriving from abroad often undergo additional customs and immigration procedures, which can contribute to longer turnaround times.
#### 2. **Gate Unavailability** – If arriving aircraft can’t access a free gate, they may need to wait on the tarmac, even after landing.
</span>

In [36]:
top_10_performances_arrival = origin_data_2.pivot_table(index='DESTINATION_AIRPORT', values='ARRIVAL_DELAY', aggfunc='mean').sort_values(by='ARRIVAL_DELAY', ascending=False).reset_index()[-10:]
top_10_performances_arrival

DESTINATION_AIRPORT  ARRIVAL_DELAY
268               12478          -19.0
269                 APN          -19.0
270               13232          -20.0
271               11471          -20.0
272                 LBE          -22.0
273                 ESC          -22.0
274                 CAK          -23.0
275                 SUX          -24.0
276               12339          -26.5
277               13290          -31.0

## <span style="color: green"> **What Could Be Further Investigated?** </span>
### 1. **Most Delay-Prone Routes**
#### Which routes are most affected by delays?
#### Are the delays primarily due to weather conditions or are they caused by airport-specific inefficiencies (e.g., congestion, limited runways, staff shortages)?
#### Understanding the root cause per route could help in recommending operational or scheduling improvements.

### 2. **Inconsistent Destination Airport IDs**
#### Why do some destination airports appear with numeric IDs instead of standard IATA letter codes?
#### This inconsistency could hinder quick identification, especially in data visualization tools and dashboards.
#### A data cleaning or mapping process may be necessary for accurate reporting and user experience.


### 3. **Aircraft-Specific Delay Patterns**
#### Do certain aircraft models experience more delays due to:
##### - Age (older planes requiring longer maintenance)?
##### - Technology gaps (e.g., outdated navigation or communication systems)?
##### - Manufacturer-specific trends (e.g., some models prone to technical issues)?
##### Investigating this could reveal performance inefficiencies and help optimize fleet usage.

-----------------------

# <span style="color: green"> **Next steps** </span>

## Expand the Dataset for Predictive Modeling
### To enable effective forecasting of delays and inefficiencies, the next step would be to source a larger and more comprehensive dataset that allows for a train/test split and robust model training.

### **Key Data Enhancements to Look For**:
#### <span style="color: green"> 1.Aircraft Details:</span>
##### - Year of construction
##### - Engine type
##### - Fuel capacity
#### <span style="color: green"> 2. Geospatial Information:</span>
##### - Departure and arrival airports (for mapping and regional trends)
##### - Mapping and Regional Trends: Helps to analyze how different airports perform under various conditions.
##### <span style="color: green"> 3. Temporal Features:
##### - Time of day
##### - Seasonality (month, holiday periods, etc.)
#### <span style="color: green"> 4. Airline Characteristics:</span>
##### - Airline identity
##### - Historical performance (track record of delays or on-time rates)
#### <span style="color: green"> 5. External Data:</span>
##### - Weather conditions
##### - Airspace restrictions (temporary closures, military exercises, etc.)
##### - Strikes or labor disputes (e.g., air traffic controllers, ground staff, pilots)
##### - Large-scale events (sporting events, festivals, expos)

### <span style="color: green"> **Reinforcement Learning** </span>
#### Reinforcement learning (RL) can be explored to improve predictive modeling for delays. By using RL, the system can learn optimal policies for operational decisions such as:
#### - When to delay or reschedule flights based on real-time conditions
#### - Dynamic pricing for compensation based on delay duration and passenger impact
#### - Adjustments for operational inefficiencies that cause delays
#### RL can be particularly effective in environments with changing conditions and uncertain outcomes, like predicting delays caused by weather, staffing, or airspace issues.

##### With these additional features, we could train more accurate models to predict delays, identify inefficiency patterns, and ultimately provide actionable insights for operational improvement.

